src/plan/global.rs

```diff
@@ -678,7 +693,9 @@ impl<VM: VMBinding> BasePlan<VM> {
             .store(emergency_collection, Ordering::Relaxed);
 
         if emergency_collection {
-            plan.force_full_heap_collection();
+            if let Some(gen) = plan.generational() {
+                gen.force_full_heap_collection();
```

Comment:
it's possible to call `plan.force_full_heap_collection();` all the time. For non-generational full-heap collectors this is just a no-op.

Reply:
That is what the current code does, and `force_full_heap_collection()` is a method of `trait Plan`. I moved the method to `trait GenerationalPlan`. I think it is cleaner, as not all plans are generational, and the idea of full heap does not apply to them.